File: conformance/tests/stateful/stateful.runar.ts
import { StatefulSmartContract, assert } from 'runar-lang';

class Stateful extends StatefulSmartContract {
  count: bigint;
  readonly maxCount: bigint;

  constructor(count: bigint, maxCount: bigint) {
    super(count, maxCount);
    this.count = count;
    this.maxCount = maxCount;
  }

  public increment(amount: bigint): void {
    this.count = this.count + amount;
    assert(this.count <= this.maxCount);
  }

  public reset(): void {
    this.count = 0n;
  }
}

File: conformance/tests/stateful/stateful.runar.rs
use runar::prelude::*;

#[runar::contract]
struct Stateful {
    count: Int,
    #[readonly]
    max_count: Int,
}

#[runar::methods(Stateful)]
impl Stateful {
    #[public]
    fn increment(&mut self, amount: Int) {
        self.count = self.count + amount;
        assert!(self.count <= self.max_count);
    }

    #[public]
    fn reset(&mut self) {
        self.count = 0;
    }
}
